### PR TITLE
perf: use one RPC call on getData multiple

### LINKED ISSUE #49: Use only one RPC call for `getAllData()`
### Summary

If lib wants to request multiple data keys: A,B, C
it will perform multiple: `getData()` calls:

- For legacy: `getData(A)`, `getData(B)`, `getData(C)`
- For new: `getData([A])`, `getData([B])`, `getData([C])`

It should instead send only one request:

- For legacy: `getDataMultiple([A, B, C])`
- For new: `getData([A, B, C])`

### PR DESCRIPTION
**Pull request by Hugoo**
### What kind of change does this PR introduce (bug fix, feature, docs update, ...)?

Performances

### What is the current behaviour (you can also link to an open issue here)?

See issue #49 

### What is the new behaviour (if this is a feature change)?

Will use only 1 RPC call when possible

### Other information:

Closes #49 

Screenshot 2021-11-24 at 09 50 05

### CODE REVIEW

File: test/mockProviders.ts [hunk 1]
```diff
@@ -181,18 +199,27 @@ export class EthereumProvider {
         {
           const keyParam = '0x' + payload.params[0].data.substr(10);
 
-          result = Array.isArray(this.returnData)
-            ? this.returnData.find((e) => e.key === keyParam).value
-            : this.returnData;
+          result = this.returnData.find((e) => e.key === keyParam).value;
```
Comment (rryter) [inline test/mockProviders.ts:202]: can `this.returnData.find((e) => e.key === keyParam)` ever return undefined?

**Review by rryter — APPROVED** [review]

**Review by CallumGrindle — APPROVED** [review]

File: src/providers/ethereumProviderWrapper.ts [hunk 2]
```diff
@@ -125,21 +130,65 @@ export class EthereumProviderWrapper {
...
+      case ERC725_VERSION.ERC725:
+        return this.getAllDataNonLegacy(address, keys);
+      case ERC725_VERSION.ERC725_LEGACY:
+        return this.getAllDataLegacy(address, keys);
+      default:
+        return [];
+    }
+  }
+
+  private async getAllDataNonLegacy(
```
Comment (frozeman) [inline src/providers/ethereumProviderWrapper.ts:152]: what a horrible name :)

`getData` should be it and if there is OLD function you call it `getDataLegacy`

**Comment by frozeman** [comment]
for keytype array we can only do it in 2 calls, as we first need to get the length.. Is that considered?

File: src/providers/ethereumProviderWrapper.ts (hunk 2 above)
Comment (Hugoo) [reply src/providers/ethereumProviderWrapper.ts:152]: There is already a function named getData, it was the one used in previous code as there was no distinction between legacy and no legacy. I didn't know what else I could use as getData was already taken.... If I call this one getData, how do I call the other one then?

**Comment by Hugoo** [comment]
> for keytype array we can only do it in 2 calls, as we first need to get the length.. Is that considered?

Yes this is handled in another part of the code, this update won't affect this.

File: test/mockProviders.ts (hunk 1 above)
Comment (Hugoo) [reply test/mockProviders.ts:202]: Good catch, yes it potentially can.
So far we only use it with a set of data we know will always be true as the mock input and "expected return data" come from same object. But adding a `?` can't hurt, I'll update thanks.

File: src/providers/ethereumProviderWrapper.ts (hunk 2 above)
Comment (Hugoo) [reply src/providers/ethereumProviderWrapper.ts:152]: I'm thinking maybe the current "getData" function can be removed at least for new ERC725Y as the command is the same, only the number of items in the array is different. I will check this tomorrow and make it more "lean". There might be some useless logic duplication

**Review by frozeman — APPROVED** [review]

**Pull request merged**